jenkins/cibuilder.groovy

```diff
@@ -0,0 +1,124 @@
+/*
+ *  Utility for deploying parallel pipeline builds in CI.
+ *
+ *  Copyright (C) 2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+/* Windows slave labels. Used for checking if target platform is Windows. */
+windows_labels = [
+    "windows-tls"
+]
+
+/**
+ * \brief   Create a sub job for given test and platform.
+ *
+ * \param test_name Descriptive test name. Like: cmake-asan-debian-x64
+ * \param platform  Test platform.
+ * \param docker_lbl
+ *                  Docker label if platform is on a docker container. Else null.
+ * \param src_stash_name
+ *                  Stash name for unstashing source code.
+ *
+ * \return          Returns a node block that is run in parallel by the caller.
+ */
+def create_subjob( test_name, platform, docker_lbl, src_stash_name ) {
+    if( docker_lbl ) {
+        return {
+            node( docker_lbl ) {
+                timestamps {
+                    deleteDir()
+                    unstash src_stash_name
+                    sh """
```

Comment:
All `sh` scripts should start with `set -e`. (Unless this is implicit? But IIRC we've had this problem before: it isn't.)

Reply:
For ```sh``` ```set -e``` behaviour is implicit. But for robustness there is no harm in adding this. For ```batch``` files there is no equivalent option. Hence need to check error and exit.